### PR TITLE
Upload defaultpath file only once

### PR DESCRIPTION
**Pull request by janos**
This PR changes how uploading a directory with a default entry works.

Currently, the file that is specified as the default entry on upload is uploaded twice if it is inside the uploading directory or any subdirectory. This is results in unnecessary processing on uploading. It creates another issue with encrypted uploads, as the same content is uploaded twice resulting the different hashes in default entry and entry that has the same path in the manifest. Beside unnecessary processing and storage use, there is a problem to identify which file in manifest is used as the default entry, as only hash can be used for that. Using cli command `manifest update` on the same file path that is set for the default entry is expected to update the default manifest entry as well. But, in case of encrypted content, hashes are not the same as the file is uploaded twice, and it is not possible to identify if the updated file is used as the default entry in upload.

This PR changes the API, client and cli in a breaking way.

- API and client UploadTar function accepts argument defaultPath.
- HTTP handleTarUpload now has an optional query parameter `defaultpath`, named the same as the cli flag, with a value to a path inside the tar archive to be set as the default entry.

Features:

- CLI `up --recursive` now allows --defaultpath value to be the relative path inside the uploading dir as well as an absolute one.

Adding default path option to tar upload, double uploading is prevented and manifests for encrypted uploads have the same hash in default entry and the one with the same content inside the manifest or any nested manifest as well.

One disadvantage is that default file can not be outside of uploading dir on cli recursive upload.

NOTE: `manifest update` functionality is addressed in a separate PR https://github.com/ethersphere/go-ethereum/pull/812 where this issue has been discovered.

### CODE REVIEW

File: swarm/api/client/client.go [hunk 1]
```diff
@@ -175,7 +175,10 @@ func (c *Client) UploadDirectory(dir, defaultPath, manifest string, toEncrypt bo
 	} else if !stat.IsDir() {
 		return "", fmt.Errorf("not a directory: %s", dir)
 	}
-	return c.TarUpload(manifest, &DirectoryUploader{dir, defaultPath}, toEncrypt)
+	if _, err := os.Stat(filepath.Join(dir, defaultPath)); err != nil {
```
Comment (lmars) [inline swarm/api/client/client.go]: Should this only be done if `defaultPath` is set?

File: swarm/api/client/client.go [hunk 2]
```diff
@@ -175,7 +175,10 @@ func (c *Client) UploadDirectory(dir, defaultPath, manifest string, toEncrypt bo
...
+		return "", fmt.Errorf("default path: %v", err)
```
Comment (lmars) [inline swarm/api/client/client.go]: This should probably be more informative in the typical case of `os.IsNotExist(err)`, e.g. `the default path "index.html" was not found in the upload directory "/path/to/dir"`

File: swarm/api/client/client.go (hunk 1 above)
Comment (janos) [reply swarm/api/client/client.go]: Thanks, that is correct.

**Comment by cobordism** [comment]
Can --recursive up also work with an absolute path given to the --defaultpath as long as that absolute path happens to be a file within the same directory being recursively uploaded?

This would keep current documentation examples and videos working. The only examples we have at the moment are of the form
```
/path/to/a $> swarm --defaultpath /path/to/a/defaultfile --recursive up ./
```
i.e. `/path/to/a/defaultfile` is in the directory `./`

**Comment by janos** [comment]
@homotopycolimit will be added.

**Review by cobordism — COMMENTED** [review]
I have a problem. I did the following:
```
~/go/src/github.com/ethereum/go-ethereum$ ./build/bin/swarm  --bzzapi http://swarm-gateways.net --defaultpath /home/user/Documents/swarmstuff/www/swarm-home-upload/index.html --recursive up /home/user/Documents/swarmstuff/www/swarm-home-upload/ --encrypt
4ef87bd44a28deef608550149afa5fce4ef4c1d2ec217f32b13ebb2f74be81191a04cec156de88a864ea6c8c00ae04eea8e01fc3c7f16cc41f6c40223353eb51
```

but the resulting manifest has no default entry at all. https://swarm-gateways.net/bzz-raw:/4ef87bd44a28deef608550149afa5fce4ef4c1d2ec217f32b13ebb2f74be81191a04cec156de88a864ea6c8c00ae04eea8e01fc3c7f16cc41f6c40223353eb51/?content_type=text

is this an error in the code, or is this because the gateway is running a different version?

**Comment by janos** [comment]
@homotopycolimit Yes, you are right, node on bzzapi address must be from this branch to be able to set default entry. It is one of the breaking changes, as default entry was uploaded by the client (cli), and now it is set by swarm node based on defaultpath query parameter on tar upload post request. Could you try the same but on a local node built from code on this branch?

**Review by cobordism — APPROVED** [review]
seems to work as advertised :)

**Comment by janos** [comment]
Resubmitted to https://github.com/ethereum/go-ethereum/pull/17375 in combination with https://github.com/ethersphere/go-ethereum/pull/812.

**Pull request closed without merging**